Services/source/dfgm/dfgm_service.c

```diff
@@ -0,0 +1,193 @@
+        ex2_log("FAILED TO CREATE TASK start_dfgm_service\n");
+        return SATR_ERROR;
+    }
+    ex2_register(svc_tsk, svc_funcs);
+    ex2_log("DFGM service started\n");
+    return SATR_OK;
+}
+
+/**
+ * @brief
+ *      Takes a CSP packet and switches based on the subservice command
+ * @details
+ *      Reads/Writes data from DFGM EHs using subservices
+ * @param *packet
+ *      The CSP packet
+ * @return SAT_returnState
+ *      Success or failure
+ */
+SAT_returnState dfgm_service_app(csp_packet_t *packet) {
+    uint8_t ser_subtype = (uint8_t)packet->data[SUBSERVICE_BYTE];
+    int8_t status;
+    SAT_returnState return_state = SATR_OK; // OK until an error is encountered
+    int32_t givenRuntime = 0;
+    int32_t maxRuntime = INT_MAX; // INT_MAX = 2^31 - 1 seconds = ~68.05 yrs
+
+    switch (ser_subtype) {
+    case DFGM_RUN: {
+        // Get runtime
+        cnv8_32(&packet->data[IN_DATA_BYTE], &givenRuntime);
+
+        // Give runtime (in seconds) to the DFGM Rx Task
+        status = HAL_DFGM_run(givenRuntime);
+
+        // Return success report
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        set_packet_length(packet, sizeof(int8_t) + 1);
+        break;
+    }
+
+    case DFGM_START: {
+        // Give the max runtime (in seconds) to the DFGM Rx Task
+        status = HAL_DFGM_run(maxRuntime);
+
+        // Return success report
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        set_packet_length(packet, sizeof(int8_t) + 1);
+        break;
+    }
+
+    case DFGM_STOP: {
+        // Tell the DFGM Rx Task to stop running
+        status = HAL_DFGM_stop();
+
+        // Return success report
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        set_packet_length(packet, sizeof(int8_t) + 1);
+        break;
+    }
+
+    case DFGM_GET_HK: {
+        // Get DFGM HK data
+        DFGM_Housekeeping HK = {0};
+        status = HAL_DFGM_get_HK(&HK);
+
+        // Convert floats from host byte order to server byte order
+        HK.coreVoltage = csp_hton16(HK.coreVoltage);
+        HK.sensorTemp = csp_hton16(HK.sensorTemp);
+        HK.refTemp = csp_hton16(HK.refTemp);
+        HK.boardTemp = csp_hton16(HK.boardTemp);
+        HK.posRailVoltage = csp_hton16(HK.posRailVoltage);
+        HK.inputVoltage = csp_hton16(HK.inputVoltage);
+        HK.refVoltage = csp_hton16(HK.refVoltage);
+        HK.inputCurrent = csp_hton16(HK.inputCurrent);
+        HK.reserved1 = csp_hton16(HK.reserved1);
+        HK.reserved2 = csp_hton16(HK.reserved2);
+        HK.reserved3 = csp_hton16(HK.reserved3);
+        HK.reserved4 = csp_hton16(HK.reserved4);
+
+        // Return success report and DFGM HK data
+        memcpy(&packet->data[STATUS_BYTE], &status, sizeof(int8_t));
+        memcpy(&packet->data[OUT_DATA_BYTE], &HK, sizeof(HK));
+        set_packet_length(packet, sizeof(int8_t) + sizeof(HK) + 1);
+        break;
```

Comment:
Yeah I've ran the service code before and confirmed all of it works. Last time I ran it though, the HK values were floats instead of uint16, but I don't think that a change in HK data types should break anything considering that I already changed csp_htonflt() into csp_hton16()

Reply:
Everything still builds/compiles correctly btw!